research/condist-fl/README.md

```diff
@@ -9,7 +9,105 @@ Accepted to the 4th Workshop on Distributed, Collaborative, & Federated Learning
 
 ## License
-TBD
+The code in this directory is released under Apache v2 License.
+
+## System Requirements
+
+The sample code is developed and tested on DGX-Station V100 with
+ - NVIDIA V100 32GB * 4
+ - 256GB system memory
+
+Minimam requirements:
+ - 4 CUDA Capable GPUs with > 32GB memory
+ - 96GB system memory
+ - 100GB disk space
+
+## Data Preparation
+Download the liver, pancreas and spleen datasets from the [MSD Challenge](http://medicaldecathlon.com/dataaws/) and the kidney dataset from [KiTS19 Challenge](https://kits19.grand-challenge.org/data/). Once the data is downloaded, use `prepare_data.py` to convert the datasets.
+
+```
+python prepare_data.py --task kidney   --src KiTS19          --dst data/KiTS19
+python prepare_data.py --task liver    --src Task03_Liver    --dst data/Liver
+python prepare_data.py --task pancreas --src Task07_Pancreas --dst data/Pancreas
+python prepare_data.py --task spleen   --src Task09_Spleen   --dst data/Spleen
+```
+
+## Setup
+
+### Software Enviroment
+
+Install all dependencies using
+```
+pip install -r requirements.txt
+```
+
+### Configure paths for datasets
+
+For each job in `jobs` folder, the four clients `kidney`, `liver`, `pancreas`, and `spleen` all requires different datasets. The setting of the path is available in `config/config_data.json`. The following example is the `config_data.json` for the `kidney` client.
+```
+{
+  "data_root": "/work/pancreas/nvidia-tlt/scripts/condist_fl/data/KiTS19",
+  "data_list": "/work/pancreas/nvidia-tlt/scripts/condist_fl/data/KiTS19/datalist.json",
+
+  "num_samples": 2,
+
+  "dataset": {
+    "use_cache_dataset": true,
+    "cache_rate": 1.0,
+    "num_worker": 4
+  },
+
+  "data_loader": {
+    "batch_size": 2,
+    "num_workers": 4
+  }
+}
+```
+The user must change the `data_root` and `data_list` to the real path on their system.
+
+## Run
+
+### Training
+
+To run ConDistFL with default settings, use `start_fl.sh` directory. The training logs and results will be saved in `workspace`, user can use TensorBoard to monitor the training.
+```
+tensorboard --logdir workspace/simulate_jobs
+```
+
+### Validation
+
+The training process runs cross site validation (on the test set) automatically when the training ends. If the user intend to run cross site validation manually, use `run_validate.py` as following.
+```
+python run_validate.py -w workspace -o cross_site_validate.json
+```
+
+### Inference
+
+The inference script requires torchscript format model to work. User can use `convert_to_torchscript.py` to convert existing checkpoints.
+```
+python convert_to_torchscript.py --workspace workspace --app server --output best_global_model.pt
+```
+
+Once the torchscript model is available, use `run_infer.py` to run inference. The following example is how to run inference using best global model on the liver test set.
+```
+python run_infer.py \
```

Comment:
I'm getting this error trying to run this (also with the ckpt downloaded from gdrive link below)
```
python run_infer.py   --data_root data/Liver   --data_list data/Liver/datalist.json   --data_list_key testing   --model best_global_model.pt   --output infer
Traceback (most recent call last):
  File "run_infer.py", line 134, in <module>
    main(args)
  File "run_infer.py", line 116, in main
    with (torch.no_grad(), torch.cuda.amp.autocast()):
AttributeError: __enter__
```

Reply:
It seems the syntax `with (torch.no_grad(), torch.cuda.amp.autocast()):` is not supported in Python 3.8. Fixed.